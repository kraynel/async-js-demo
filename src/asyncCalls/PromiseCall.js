import React, { Component } from 'react';
import UpsideDownPromise from '../services/UpsideDownPromise';
import ButtonWithStatus from './ButtonWithStatus';

export default class PromiseCall extends Component {
  constructor(props) {
    super(props);
    this.upsideDownPromise = new UpsideDownPromise();
    this.state = {
      strangerStatus: 'In real world'
    };
  }

  toggleCall = () => {
    this.upsideDownPromise
      .enter()
      .then(() => {
        this.setState({ strangerStatus: 'In upside down' });
        return this.upsideDownPromise.findDemogorgon();
      })
      .then(location => {
        this.setState({ strangerStatus: `Demogorgon in ${location}` });
        return this.upsideDownPromise.killDemogorgon();
      })
      .then(result => {
        if (result === 'SUCCESS') {
          this.setState({ strangerStatus: 'Demogorgon dead ️☠️' });
          return;
        }

        this.setState({ strangerStatus: 'You are dead ☠️' });
      });
  };

  render() {
    return (
      <ButtonWithStatus
        toggleCall={this.toggleCall}
        buttonText="Promise Call"
        strangerStatus={this.state.strangerStatus}
      />
    );
  }
}
