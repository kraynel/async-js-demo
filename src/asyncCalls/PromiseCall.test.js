import React from 'react';
import renderer from 'react-test-renderer';

import PromiseCall from './PromiseCall';

describe('PromiseCall', () => {
  it('should start in the real world', () => {
    const component = renderer.create(<PromiseCall />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
