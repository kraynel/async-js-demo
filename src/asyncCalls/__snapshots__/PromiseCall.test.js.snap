// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PromiseCall should fail to kill the demogorgon 1`] = `
<div
  className="buttonWithStatus"
>
  <button
    className="buttonToggleUpsideDown"
    onClick={[Function]}
  >
    Promise Call
  </button>
  <span>
    You are dead ☠️
  </span>
</div>
`;

exports[`PromiseCall should find the demogorgon 1`] = `
<div
  className="buttonWithStatus"
>
  <button
    className="buttonToggleUpsideDown"
    onClick={[Function]}
  >
    Promise Call
  </button>
  <span>
    Demogorgon in SOME PLACE
  </span>
</div>
`;

exports[`PromiseCall should go in the upside down 1`] = `
<div
  className="buttonWithStatus"
>
  <button
    className="buttonToggleUpsideDown"
    onClick={[Function]}
  >
    Promise Call
  </button>
  <span>
    In upside down
  </span>
</div>
`;

exports[`PromiseCall should kill the demogorgon 1`] = `
<div
  className="buttonWithStatus"
>
  <button
    className="buttonToggleUpsideDown"
    onClick={[Function]}
  >
    Promise Call
  </button>
  <span>
    Demogorgon dead ️☠️
  </span>
</div>
`;

exports[`PromiseCall should start in the real world 1`] = `
<div
  className="buttonWithStatus"
>
  <button
    className="buttonToggleUpsideDown"
    onClick={[Function]}
  >
    Promise Call
  </button>
  <span>
    In real world
  </span>
</div>
`;
